### ISSUE TITLE
about sparse_norm_adj

### ISSUE THREAD

**Issue by Doctor-JZL** [Coder-Yu/SELFRec#12]
ego_embeddings = torch.sparse.mm(self.sparse_norm_adj, ego_embeddings)
你好，请问对于上述代码中的sparse_norm_adj，这一稀疏矩阵生成的依据是什么？做了实验，发现这个矩阵很重要。

**Comment by Coder-Yu** [comment]
这个是图神经网络的基础，你找一下相关教程吧。稀疏表示是为了降低存储空间和计算复杂度

**Comment by Doctor-JZL** [comment]
不好意思，我想问的是，你为什么这样会这样生成sparse_norm_adj这个矩阵，进而生成后序的sign信号，是与user/item 的权重有关吗？

**Comment by Coder-Yu** [comment]
没明白这里有什么值得问的问题

**Comment by Doctor-JZL** [comment]
呵呵，或者换一种说法吧，sparse_norm_adj这个矩阵用这种方式生成吗，有没有其他方式来生成它？

**Comment by Coder-Yu** [comment]
这行是torch自带的稀疏矩阵乘法的api。我真不明白这里为什么会有疑问。

如果你想问的是稀疏矩阵怎么生成的，你可能说的是这行，
self.sparse_norm_adj = TorchGraphInterface.convert_sparse_mat_to_tensor(self.norm_adj)
稀疏矩阵生成的方法有很多种，具体你可以参考torch和numpy稀疏矩阵部分的文档。

如果你想问的是用户商品的正则化矩阵是怎么生成的，建议你参考一下neural graph collaborative filtering里面的相关部分。

这位老兄，如果你真想好好问问题，建议你首先了解一下相关基础，然后把问题描述清楚而不是以非常不明确的方式提问，这样别人没法准确回答你。

**Comment by Doctor-JZL** [comment]
不好意思，确实是我表达的不够清楚，请您谅解。也非常感谢您的再次回复。请看下文，我最终想问的是self.__create_sparse_bipartite_adjacency()这个方法里adj_mat生成的问题。再次对我未能表达清楚我的问题对您道歉。
ego_embeddings = torch.sparse.mm(self.sparse_norm_adj, ego_embeddings) <- self.sparse_norm_adj = TorchGraphInterface.convert_sparse_mat_to_tensor(self.norm_adj).cuda() <- self.norm_adj = self.normalize_graph_mat(self.ui_adj) <- self.ui_adj = self.__create_sparse_bipartite_adjacency()

**Comment by Doctor-JZL** [comment]
不好意思，确实是我表达的不够清楚，请您谅解。也非常感谢您的再次回复。请看下文，我最终想问的是self.__create_sparse_bipartite_adjacency()这个方法里adj_mat生成的问题。再次对我未能表达清楚我的问题对您道歉。
ego_embeddings = torch.sparse.mm(self.sparse_norm_adj, ego_embeddings) 
<- self.sparse_norm_adj = TorchGraphInterface.convert_sparse_mat_to_tensor(self.norm_adj).cuda() 
<- self.norm_adj = self.normalize_graph_mat(self.ui_adj) 
<- self.ui_adj = self.__create_sparse_bipartite_adjacency()

    def __create_sparse_bipartite_adjacency(self, self_connection=False):
        '''
        return a sparse adjacency matrix with the shape (user number + item number, user number + item number)
        '''
        n_nodes = self.user_num + self.item_num
        row_idx = [self.user[pair[0]] for pair in self.training_data]
        col_idx = [self.item[pair[1]] for pair in self.training_data]
        user_np = np.array(row_idx)
        item_np = np.array(col_idx)
        ratings = np.ones_like(user_np, dtype=np.float32)
        tmp_adj = sp.csr_matrix((ratings, (user_np, item_np + self.user_num)), shape=(n_nodes, n_nodes),dtype=np.float32)
        adj_mat = tmp_adj + tmp_adj.T
        if self_connection:
            adj_mat += sp.eye(n_nodes)
        return adj_mat

**Comment by Coder-Yu** [comment]
![image](https://user-images.githubusercontent.com/6328763/196425298-e7c208df-bc2c-4db0-948e-6c0a9e5c3529.png)
建议你参考下neural graph collaborative filtering这篇文章

**Comment by Doctor-JZL** [comment]
好的，非常感谢，以后我会提高描述问题的技术的，感谢！
